Improve video player controls and added new functionalities

vinothvino42 opened the pull request:
Ticket: #1049

### Getters
1. isInitialized
2. isBuffering
3. isLooping
4. isPlaying
5. isEnded

### Setters
1. enableLoop property
2. autoplay property
3. volume property (0 to 1.0)
4. playbackRate property (0 to 2.0)
5. showControls property
6. loadingWidget property

### Methods

1. play method
2. pause method
3. stop method
4. mute method
5. unmute method
6. seekTo method

### Callbacks
1. onStart callback
2. onEnd callback
3. onChange callback (with position, milliseconds)

### Others
1. Added backward and forward with 10 seconds in the default controls

Sample EDL:

```yaml
View:
  header:
    title: Video
  styles:
    scrollableView: true
  body:
    Column:
      styles: { gap: 16, padding: 24, crossAxis: center }
      children:
        - Text:
            id: onChangeId
            text: onChange
        - Video:
            id: videoId
            # source: https://flutter.github.io/assets-for-api-docs/assets/videos/bee.mp4
            # source: http://commondatastorage.googleapis.com/gtv-videos-bucket/sample/BigBuckBunny.mp4
            source: https://storage.googleapis.com/gtv-videos-bucket/sample/BigBuckBunny.mp4
            loadingWidget: VideoLoader
            autoplay: false
            repeat: false
            playbackRate: 1
            volume: 1
            onStart: |
              //@code
              console.log('Started');
            onChange: |
              //@code
              console.log('onChange called ' + event.data.position);
              onChangeId.text = 'onChange called: Position(MS) - ' +event.data.position;
            onEnd: |
              //@code
              console.log('Ended');

        - Row:
            styles: { gap: 16, mainAxis: center }
            children:
              - Button:
                  label: Play
                  onTap: |
                    //@code
                    videoId.play();

              - Button:
                  label: Pause
                  onTap: |
                    //@code
                    videoId.pause();

              - Button:
                  label: Stop
                  onTap: |
                    //@code
                    videoId.stop();

        - Row:
            styles: { gap: 16, mainAxis: center }
            children:
              - Button:
                  label: Mute
                  onTap: |
                    //@code
                    videoId.mute();

              - Button:
                  label: Unmute
                  onTap: |
                    //@code
                    videoId.unmute();

              - Button:
                  label: Seek
                  onTap: |
                    //@code
                    videoId.seekTo(120);

        - Column:
            crossAxis: center
            styles:
              gap: 10
            children:
              - Text:
                  text: PlaybackRate
              - Row:
                  styles: { gap: 16, mainAxis: center }
                  children:
                    - Button:
                        label: Fast
                        onTap: |
                          //@code
                          videoId.playbackRate = 2.0;

                    - Button:
                        label: Slow
                        onTap: |
                          //@code
                          videoId.playbackRate = 0.5;

        - Column:
            crossAxis: center
            styles:
              gap: 10
            children:
              - Text:
                  text: Volume
              - Row:
                  styles: { gap: 16, mainAxis: center }
                  children:
                    - Button:
                        label: Set Vol (0.3)
                        onTap: |
                          //@code
                          videoId.volume = 0.3;

                    - Button:
                        label: Set Vol (1.0)
                        onTap: |
                          //@code
                          videoId.volume = 1.0;

VideoLoader:
  LoadingContainer:
    id: loadingWidget
    isLoading: true
    useShimmer: true
    defaultShimmerPadding: 0 20
    loadingWidget:
      Shape:
        styles:
          height: 200
          width: 200
          borderRadius: 8
          backgroundColor: 0xFFF5F5F7

```

vusters commented on lib/widget/video.dart:
```diff
@@ -23,42 +26,102 @@ class Video extends StatefulWidget
+          _controller.showControls = Utils.getBool(value, fallback: true),
+      'loadingWidget': (loadingWidget) =>
+          _controller.loadingWidget = loadingWidget,
+      'enableLoop': (value) => _controller._playerController
```
use "repeat" instead?

vinothvino42 replied:
Okay I'll update

vusters commented on lib/widget/video.dart:
```diff
@@ -23,42 +26,102 @@ class Video extends StatefulWidget
+      'mute': () => _controller.playerCapabilities?.mute(),
+      'unmute': () => _controller.playerCapabilities?.unmute(),
+      'seekTo': (value) => _controller.playerCapabilities?.seekTo(value),
+      'setPlaybackRate': (value) =>
```
do we need this if we already have the setter?

vinothvino42 replied:
Updated

vusters commented on assets/schema/ensemble_schema.json:
```diff
@@ -6569,6 +6569,42 @@
+        },
+        "onChange": {
+          "$ref": "#/$defs/Action-payload",
+          "description": "It is called everytime when the player is playing the video"
```
refine this. Every time a "new" video is played?

vusters commented on lib/widget/video.dart:
```diff
@@ -23,42 +26,102 @@ class Video extends StatefulWidget
+  EnsembleAction? onStart;
+  EnsembleAction? onChange;
+  EnsembleAction? onEnd;
+  EnsembleAction? onComplete;
```
not exposed in schema.

vinothvino42 replied:
onComplete is not used in the code. Just removed it.

kmahmood74 reviewed (commented):
@vinothvino42 please test with multiple video players on the same screen. At least 3 and make sure you are able to control each one separately. Create a test page for that with controls that are triggered with buttons. See the Youtube example in kitchen sink.

kmahmood74 commented on lib/widget/video.dart:
```diff
@@ -77,41 +141,104 @@ class MyController extends WidgetController {
+class VideoState extends WidgetState<Video> with PlayerCapabilities {
+  @override
+  void didChangeDependencies() {
+    widget._controller.playerCapabilities = this;
```
this can cause memory leak as you are not setting to null in dispose of the state as a result the last state can hang around in memory. 

widget._controller.playerCapabilities = null; inside dispose() should fix it

vinothvino42 replied:
Done

vinothvino42 commented:
@vusters 
New Changes:

1. Changed the property name from `enableLoop` to `repeat`
2. Removed the **setPlaybackRate** and **setVolume** method and now it can be changed using the setter alone
3. Added playCapabilities = null inside the dispose method
4. Removed onComplete callback
5. Updated schema
